src/librustc_mir/interpret/validity.rs

```diff
@@ -580,10 +580,13 @@ impl<'rt, 'mir, 'tcx, M: Machine<'mir, 'tcx>> ValueVisitor<'mir, 'tcx, M>
             }
             ty::Array(tys, ..) | ty::Slice(tys)
                 if {
-                    // This optimization applies only for integer and floating point types
-                    // (i.e., types that can hold arbitrary bytes).
+                    // This optimization applies for types that can hold arbitrary bytes (such as
+                    // integer and floating point types) or for types that are empty (such as the
```

Comment:
These types are not empty. `!` is an empty type, it has no inhabitants. We can certainly not skip validating those.

The types you are skipping here are inhabited ZST -- they are all equivalent to `()`, the unit type.

Reply:
Yes, that's a much more accurate way of describing this. Thank you!